lib/stitches/api_key.rb

```diff
@@ -27,13 +27,7 @@ def do_call(env)
       if authorization
         if authorization =~ /#{@configuration.custom_http_auth_scheme}\s+key=(.*)\s*$/
           key = $1
-
-          if ApiClient.column_names.include?("enabled")
-            client = ApiClient.where(key: key, enabled: true).first
-          else
-            ActiveSupport::Deprecation.warn('api_keys is missing "enabled" column.  Run "rails g stitches:add_enabled_to_api_clients"')
-            client = ApiClient.where(key: key).first
-          end
+          client = ApiClient.fetch_for_key(key)
```

Comment:
I was looking below and wondering how to get out of the boat of modifying ApiClient ... when I did a variation of this caching earlier I modified ApiClient and while the functionality has been successful, the modification to that class was always a little bit of a surprise. This is a bit less of a concern in what you're proposing here because you've defined new cache methods as opposed to manipulating `where` like I did.

But I think we can avoid the meta-stuff entirely? How about ...

```ruby
          client = ApiClientCacheWrapper.fetch_for_key(key)
```

... then ...

```ruby
require 'lru_redux'

class Stitches::ApiClientCacheWrapper
  MAX_ITEMS_IN_CACHE = ENV.fetch("API_KEY_CACHE_MAX_ITEMS", 100)
  CACHE_DURATION_IN_SECONDS = ENV.fetch("API_KEY_CACHE_DURATION", 5 * 60) # five minutes

  API_KEY_CACHE = LruRedux::TTL::Cache.new(
    Stitches::CacheInjector::MAX_ITEMS_IN_CACHE,
    Stitches::CacheInjector::CACHE_DURATION_IN_SECONDS
  )

  def clear_api_cache
    API_KEY_CACHE.clear
  end

  def fetch_for_key
    if ::ApiClient.column_names.include?("enabled")
      API_KEY_CACHE.getset(key) do
        ApiClient.where(key: key, enabled: true).first
      end
    else
      ActiveSupport::Deprecation.warn('api_keys is missing "enabled" column.  Run "rails g stitches:add_enabled_to_api_clients"')
      API_KEY_CACHE.getset(key) do
        ApiClient.where(key: key).first
      end
    end
  end
end
```

Reply:
Future thing, not a blocker™ I can imagine the storage getting passed in as well:

```ruby
class Stitches::ApiClientCacheWrapper
  def initialize(storage = Stitches::Configuration.config.storage) # defaults to ApiClient
    # ...
   end

  # ...
  @storage.where(key: key, enabled: true).first
  # ...
```

```ruby 
# in application.rb
config.stitiches.storage = WhateverStorage
```

If we did something like that we might think about the storage api which currently now is
`.where(key: key, enabled: true).first` ... maybe it should be `.find_by(key: key, enabled: true)`? Basically `WhateverStorage` would just need to support whatever we decided there.

Reply:
:palm-to-head:, This was such a great insight. I was so focused on extending what we do in the app I forgot that from stitches we control the calling conventions.